libs/application-generic/src/services/analytic-logs/step-run/step-run.repository.ts

```diff
@@ -0,0 +1,187 @@
+import { Injectable } from '@nestjs/common';
+import { PinoLogger } from 'nestjs-pino';
+import { JobEntity, JobStatusEnum, MessageEntity } from '@novu/dal';
+import { FeatureFlagsKeysEnum, StepTypeEnum } from '@novu/shared';
+import { addYears, format } from 'date-fns';
+import { LogRepository, SchemaKeys } from '../log.repository';
+import { ClickHouseService } from '../clickhouse.service';
+import { FeatureFlagsService } from '../../feature-flags/feature-flags.service';
+import { stepRunSchema, ORDER_BY, TABLE_NAME, StepRun, StepType } from './step-run.schema';
+
+type StepRunInsertData = Omit<StepRun, 'id'>;
+
+type StepOptions = {
+  status?: JobStatusEnum;
+  message?: MessageEntity;
+  executionDurationMs?: number;
+  errorCode?: string;
+  errorMessage?: string;
+  deferredMs?: number;
+};
+
+@Injectable()
+export class StepRunRepository extends LogRepository<typeof stepRunSchema, StepRun> {
+  public readonly table = TABLE_NAME;
+  public readonly schema = stepRunSchema;
+  public readonly schemaOrderBy: SchemaKeys<typeof stepRunSchema>[] = ORDER_BY;
+  public readonly identifierPrefix = 'sr_';
+
+  constructor(
+    protected readonly clickhouseService: ClickHouseService,
+    protected readonly logger: PinoLogger,
+    protected readonly featureFlagsService: FeatureFlagsService
+  ) {
+    super(clickhouseService, logger, stepRunSchema, ORDER_BY, featureFlagsService);
+    this.logger.setContext(this.constructor.name);
+  }
+
+  private mapStepTypeEnumToStepType(stepType: StepTypeEnum | undefined): StepType | null {
+    switch (stepType) {
+      case StepTypeEnum.EMAIL:
+        return 'email';
+      case StepTypeEnum.SMS:
+        return 'sms';
+      case StepTypeEnum.IN_APP:
+        return 'in_app';
+      case StepTypeEnum.PUSH:
+        return 'push';
+      case StepTypeEnum.CHAT:
+        return 'chat';
+      case StepTypeEnum.DIGEST:
+        return 'digest';
+      case StepTypeEnum.TRIGGER:
+        return 'trigger';
+      case StepTypeEnum.DELAY:
+        return 'delay';
+      case StepTypeEnum.CUSTOM:
+        return 'custom';
+      default:
+        return null;
+    }
+  }
+
+  async create(job: JobEntity, options: StepOptions = {}): Promise<void> {
+    try {
+      const isEnabled = await this.featureFlagsService.getFlag({
+        key: FeatureFlagsKeysEnum.IS_STEP_RUN_LOGS_ENABLED,
+        organization: { _id: job._organizationId },
+        environment: { _id: job._environmentId },
+        user: { _id: job._userId },
+        defaultValue: false,
+      });
+
+      if (!isEnabled) {
+        return;
+      }
+
+      // Preserve existing deferredMs if not explicitly provided
+      const existingDeferredMs = await this.getExistingDeferredMs(job._organizationId, job._id);
```

Comment:
Can be removed

Reply:
we need this one, this one is preserving the delay amount in ms